### PR TITLE
feat(memegenerator): Add MemeGenerator support

### PR DESCRIPTION
**Pull request by badawoll**
The code symbols were renamed before the release - the fingerprints therefore heavily rely on opcode matching (I hope this is fine?).

Currently only tested on the latest version, but I can test older versions if necessary.

### CODE REVIEW

File: src/main/kotlin/app/revanced/patches/memegenerator/misc/pro/patch/UnlockProVersionPatch.kt [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import app.revanced.patcher.patch.PatchResultSuccess
+import app.revanced.patcher.patch.annotations.DependsOn
+import app.revanced.patcher.patch.annotations.Patch
+import app.revanced.patches.memegenerator.detection.license.patch.LicenseValidationPatch
+import app.revanced.patches.memegenerator.detection.signature.patch.SignatureVerificationPatch
+import app.revanced.patches.memegenerator.misc.pro.annotations.UnlockProCompatibility
+import app.revanced.patches.memegenerator.misc.pro.fingerprint.IsFreeVersionFingerprint
+
+@Patch
+@Name("memegenerator-unlock-pro")
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/memegenerator/misc/pro/patch/UnlockProVersionPatch.kt:20]:
```suggestion
@Name("unlock-pro")
```

File: src/main/kotlin/app/revanced/patches/memegenerator/detection/license/patch/LicenseValidationPatch.kt [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+import app.revanced.patcher.annotation.Version
+import app.revanced.patcher.data.BytecodeContext
+import app.revanced.patcher.extensions.replaceInstructions
+import app.revanced.patcher.patch.BytecodePatch
+import app.revanced.patcher.patch.PatchResult
+import app.revanced.patcher.patch.PatchResultSuccess
+import app.revanced.patches.memegenerator.detection.license.annotations.DisableLicenseValidationCompatibility
+import app.revanced.patches.memegenerator.detection.license.fingerprint.LicenseValidationFingerprint
+
+@Description("Disable FireBase license validation in meme generator.")
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/memegenerator/detection/license/patch/LicenseValidationPatch.kt:14]:
```suggestion
@Description("Disable FireBase license validation.")
```

File: src/main/kotlin/app/revanced/patches/memegenerator/detection/signature/patch/SignatureVerificationPatch.kt [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import app.revanced.patcher.data.BytecodeContext
+import app.revanced.patcher.extensions.replaceInstructions
+import app.revanced.patcher.patch.BytecodePatch
+import app.revanced.patcher.patch.PatchResult
+import app.revanced.patcher.patch.PatchResultSuccess
+import app.revanced.patcher.patch.annotations.Patch
+import app.revanced.patches.memegenerator.detection.signature.annotations.DisableSignatureDetectionCompatibility
+import app.revanced.patches.memegenerator.detection.signature.fingerprint.VerifySignatureFingerprint
+
+@Description("Disables detection of incorrect signature in meme generator.")
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/memegenerator/detection/signature/patch/SignatureVerificationPatch.kt:16]:
```suggestion
@Description("Disables detection of incorrect signature.")
```

**Review by oSumAtrIX — APPROVED** [review]
Looks good. The patch is constrained to a version, therefor the fingerprints will reliably on it. Though, in case the app updates the opcode pattern might change, therefor the @FuzzyPatternScanMethod annotation can be utilised. Additionally it might be possible to create a safer fingerprint instead if it is assumed that the current one might break for future versions with the main benefit of getting rid of the version constraint. Though since the patch is constrained it is fine for now.

**Review by oSumAtrIX — APPROVED** [review]

**Comment by badawoll** [comment]
Alright, I'll see what happens in future versions and adjust the fingerprints accordingly.

**Comment by oSumAtrIX** [comment]
Please drop the chore commits to resolve merge conflicts & rebase.

**Comment by badawoll** [comment]
Done, should be ready to merge now.

**Review by oSumAtrIX — APPROVED** [review]
The patch is currently constrained to a single version. Ideally, that constraint can be removed by strengthening the compatibility of fingerprints for future versions by reducing the opcode pattern size, for example.

**Pull request merged**